### DISCUSSION TITLE
How can I set the value of the input only after validating what the user typed?

### DISCUSSION THREAD

**Discussion by TitoSniberb** [react-hook-form/react-hook-form#9232] [Q&A] [answered]
Having an input, I want to prevent the user from typing in anything but numbers or a comma.

I tried to pass an onChange function to the register method, and inside my custom handler use the `setValue` method to change the value of the input, but it is getting ignored

```
const handleNumberChange = (event, field) => {
    const value = event.target.value;
    const rightFormat =
      /^[1-9]+(,\d{0,2})?$/.test(value) ||
      /^0(,\d{0,2})?$/.test(value) ||
      /^[1-9]{1}[0-9]+(,\d{0,2})?$/.test(value) ||
      value === '';

    if (rightFormat) {
      setValue(field, value);
    }
  }

<input
            placeholder={`Desde €`}
            register={register('from', {
              required: { value: true, message: 'El campo desde es requerido' },
              onChange: (e) => handleNumberChange(e, 'from'),
            })}
            autoComplete="off"
          />

```

**Comment by Moshyfawn** [comment, marked as answer]
As soon as you need to "control" your text input value _(text formatting, for example)_, you should use controlled inputs _(`Controller` / `useContoller` API)_ and format the user value `onChange` before passing it into the RHF `onChange` handler
